### ISSUE TITLE
Function with assignment to global variable may be incorrectly tree-shaken

### ISSUE THREAD

**Issue by liuly0322** [rollup/rollup#5579]
### Rollup Version

4.18.1

### Operating System (or Browser)

Chrome

### Link To Reproduction

https://rollupjs.org/repl/?version=4.18.1&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmxldCUyMHJvb3RFbGVtZW50JTIwJTNEJTIwbnVsbCU1Q25mdW5jdGlvbiUyMGdldFJvb3RFbGVtZW50KCklMjAlN0IlNUNuJTIwJTIwcm9vdEVsZW1lbnQlMjAlM0QlMjBkb2N1bWVudC5xdWVyeVNlbGVjdG9yKCdodG1sJyklNUNuJTIwJTIwcmV0dXJuJTIwcm9vdEVsZW1lbnQlNUNuJTdEJTVDbiU1Q25mdW5jdGlvbiUyMG1vZGlmeVJvb3RFbGVtZW50KCklMjAlN0IlNUNuJTIwJTIwZ2V0Um9vdEVsZW1lbnQoKSUzRi5jbGFzc0xpc3QuYWRkKCdtb2RpZmllZCcpJTVDbiU3RCU1Q24lNUNubW9kaWZ5Um9vdEVsZW1lbnQoKSUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTJDJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlN0QlMkMlMjJ0cmVlc2hha2UlMjIlM0F0cnVlJTdEJTdE

### Expected Behaviour

Generate `modifyRootElement` output, since it add a classname to `html` element.

### Actual Behaviour

output is empty

**Comment by liuly0322** [comment]
Thanks for your hint, I find `ChainExpression` may be the point. Made a [smaller version](https://rollupjs.org/repl/?version=4.18.1&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmxldCUyMG1vZGlmaWVkJTIwJTNEJTIwZmFsc2UlNUNuJTVDbmZ1bmN0aW9uJTIwc2lkZUVmZmVjdCgpJTIwJTdCJTVDbiUyMCUyMG1vZGlmaWVkJTIwJTNEJTIwdHJ1ZSU1Q24lMjAlMjByZXR1cm4lMjBudWxsJTVDbiU3RCU1Q24lNUNuc2lkZUVmZmVjdCgpJTNGLnglNUNuYXNzZXJ0Lm9rKG1vZGlmaWVkKSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTJDJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlN0QlMkMlMjJ0cmVlc2hha2UlMjIlM0F0cnVlJTdEJTdE).

```typescript
hasEffects(context: HasEffectsContext): boolean {
	if (this.expression.isSkippedAsOptional(this)) return false;
	return this.expression.hasEffects(context);
}
```

So, the code (and the corresponding `hasEffect` call) inside "sideEffect" function is "skipped", therefore the "sideEffect" function is considered pure, and removed.

@TrickyPi Are you going to fix this as I see the "triage" label is removed? I would appreciate it a lot!

**Comment by lukastaegert** [comment]
I think there is a general bug in how this is handled for ChainExpression. If in a ChainExpression, a MemberExpression.object or CallExpression.callee is null or undefined at an optional link, then
- the value of the ChainExpression is treated as `undefined` (that is correct)
- it is considered not to have side effects.

The latter is obviously wrong as it disregards all side effects left of the optional link.
Currently I think we should
- get rid of isSkippedAsOptional
- instead change (CallExpression/MemberExpression).(hasEffects/getLiteralValueAtPath) to just check the `optional` flag

I can do that unless someone else wants to go there.

**Comment by lukastaegert** [comment]
Ok, when writing this I overlooked that the AST works from right to left while the chain works from left to right. Still, I created a tentative fix here: #5589

**Comment by TrickyPi** [comment]
> @TrickyPi Are you going to fix this as I see the "triage" label is removed? I would appreciate it a lot!

Sorry for not noticing this message earlier. If I want to handle an issue, I will add myself to the assignees. So, feel free to submit a PR for any issue if it has not been assigned.
